wagtail/admin/viewsets/model.py

```diff
@@ -89,35 +132,111 @@ def index_results_view(self):
-            form_class=self.get_form_class(),
-            index_url_name=self.get_url_name("index"),
-            add_url_name=self.get_url_name("add"),
-            edit_url_name=self.get_url_name("edit"),
-            header_icon=self.icon,
+            **self.get_add_view_kwargs(),
         )
 
     @property
     def edit_view(self):
         return self.edit_view_class.as_view(
-            model=self.model,
-            permission_policy=self.permission_policy,
-            form_class=self.get_form_class(for_update=True),
-            index_url_name=self.get_url_name("index"),
-            edit_url_name=self.get_url_name("edit"),
-            delete_url_name=self.get_url_name("delete"),
-            header_icon=self.icon,
+            **self.get_edit_view_kwargs(),
         )
 
     @property
     def delete_view(self):
         return self.delete_view_class.as_view(
-            model=self.model,
-            permission_policy=self.permission_policy,
-            index_url_name=self.get_url_name("index"),
-            delete_url_name=self.get_url_name("delete"),
-            header_icon=self.icon,
+            **self.get_delete_view_kwargs(),
+        )
+
+    def get_templates(self, action="index", fallback=""):
+        """
+        Utility function that provides a list of templates to try for a given
+        view, when the template isn't overridden by one of the template
+        attributes on the class.
+        """
+        if not self.template_prefix:
+            return [fallback]
+        templates = [
+            f"{self.template_prefix}{self.app_label}/{self.model_name}/{action}.html",
+            f"{self.template_prefix}{self.app_label}/{action}.html",
+            f"{self.template_prefix}{action}.html",
+        ]
+        if fallback:
+            templates.append(fallback)
+        return templates
+
+    @cached_property
+    def index_template_name(self):
+        """
+        A template to be used when rendering ``index_view``.
+
+        Default: if :attr:`template_prefix` is specified, an ``index.html``
+        template in the prefix directory and its ``{app_label}/{model_name}/``
+        or ``{app_label}/`` subdirectories will be used. Otherwise, the
+        ``index_view_class.template_name`` will be used.
+        """
+        return self.get_templates(
+            "index",
+            fallback=self.index_view_class.template_name,
+        )
+
+    @cached_property
+    def index_results_template_name(self):
+        """
+        A template to be used when rendering ``index_results_view``.
+
+        Default: if :attr:`template_prefix` is specified, a ``index_results.html``
+        template in the prefix directory and its ``{app_label}/{model_name}/``
+        or ``{app_label}/`` subdirectories will be used. Otherwise, the
+        ``index_view_class.results_template_name`` will be used.
+        """
+        return self.get_templates(
+            "index_results",
```

Comment:
Just like `delete.html`, not sure if we should expect an `index_results.html` or a `listing_results.html` here. It's nice to be consistent with the naming, but the generic `IndexView` currently has no `results_template_name` so it will inherit the values from `BaseListingView` which has `listing.html` and `listing_results.html` as the templates.

We could also introduce an `index_results.html` template that just inherits from `listing_results.html` and set that on the generic `IndexView`...